### PR TITLE
Refactor Filesystem Code for Clarity

### PR DESCRIPTION
**Pull request by AtariDreams**
This makes the code more readable as well as lowers operations to code paths that actually need them

### CODE REVIEW

File: stl/inc/filesystem
Comment (miscco) [inline stl/inc/filesystem:1516]: This adds needless work, as `_Position` is incremented twice

File: stl/src/filesys.cpp
Comment (miscco) [inline stl/src/filesys.cpp:73]: That comment does not add value

File: stl/src/filesys.cpp
Comment (miscco) [inline stl/src/filesys.cpp:125]: I find those changes to be less readable

File: stl/src/filesys.cpp
Comment (miscco) [inline stl/src/filesys.cpp:63]: Implicit conversion to bool are conventionally avoided

File: stl/src/filesys.cpp
Comment (miscco) [inline stl/src/filesys.cpp:232]: I think BOOL is wrong here

File: stl/src/filesys.cpp
Comment (miscco) [inline stl/src/filesys.cpp:429]: Could be const

**Review by miscco — CHANGES_REQUESTED** [review]

**Comment by StephanTLavavej** [comment]
As a maintainer, I find this PR to be concerning. Every code change is risky to some degree (even innocent-seeming changes are occasionally risky, up to and including ABI breaks), and also costs time to review. With a brief glance at the changes, I can see that there's some good stuff here (e.g. I particularly like the use of the `SYMBOLIC_LINK_FLAG_ALLOW_UNPRIVILEGED_CREATE` constant which is now available), but I have these specific concerns:

1. This is a "grab bag" of different refactorings, which is harder to review. A PR that performs only one kind of transformation (e.g. [De Morgan](https://en.wikipedia.org/wiki/De_Morgan%27s_laws) changes) is significantly less risky and easier to review - even when many sites are being changed. The dangerous thing about mixed changes is that we can nod our heads to a bunch of safe changes and miss the one scary/wrong change in the middle.
   + It can sometimes be acceptable to have one PR with several kinds of refactorings cleanly grouped into separate commits. I suppose I would accept separate commits here.
2. Substantial behavior changes (whether bug fixes or performance improvements) should not be mixed with stylistic refactorings - that is, #2767 should be addressed in an isolated PR. Occasionally, because we are human (and slightly lazy), it is okay to perform one tiny cleanup in a PR that's otherwise dedicated to fixing a bug or improving performance. However, any substantial mixing really must be avoided. (Same rationale as avoiding mixed refactorings, but even stronger.)
   + Putting behavior changes and refactorings into separate commits doesn't solve this concern, they really need to be separate PRs.
3. It would be helpful for your PR description to explain why each change is being made. For some changes, I can immediately see the clarity improvement (e.g. De Morganing "test existing files for equivalence" so that it reads "if the serial numbers are the same, and the high parts are the same, and the low parts are the same, then we're equivalent" avoids the confusing negation in the original form - great!). For others, I am unsure how they are improving consistency - e.g. some conditions are being reversed to return-on-success (e.g. `__std_fs_get_current_path()`), but others are being reversed to return-on-failure (e.g. `_Open_dir()`). Having a changelog would help me and other reviewers understand the changes. It doesn't need to be *super* detailed, it just needs to mention why each function is being changed.
   + This is also useful for "source code archaeology" - as the STL is immortal, we occasionally need to look up why things were changed 5 or 10 years ago. Having a PR description that explains each change will be helpful if we need to read it in 2027 or 2032 in our flying cars.

File: stl/src/filesys.cpp
Comment (AlexGuteniev) [inline stl/src/filesys.cpp:173]: `!= 0`

File: stl/src/filesys.cpp
Comment (AlexGuteniev) [inline stl/src/filesys.cpp:234]: These preprocessor guards can be enhanced to `^^^ / vvv` (but possibly in a separate change, see #351)

File: stl/src/filesys.cpp
Comment (AlexGuteniev) [inline stl/src/filesys.cpp:356]: I don't see why we can't compare `FILE_ID_INFO` as a whole thing with a single `memcmp`

File: stl/src/filesystem.cpp
Comment (AlexGuteniev) [inline stl/src/filesystem.cpp:228]: keep `!= 0`

File: stl/src/filesys.cpp
Comment (AlexGuteniev) [inline stl/src/filesys.cpp:378]: here as well only `memcmp` can be used

Comment (AtariDreams) [reply stl/src/filesys.cpp:378]: Not really, since we are only comparing three fields, and these are not all of them

File: stl/src/filesys.cpp
Comment (StephanTLavavej) [inline stl/src/filesys.cpp:234]: This should use `uintmax_t` for consistency with the return type and `#else` codepath.

**Review by StephanTLavavej — COMMENTED** [review]
Thanks, this looks good! I found no correctness issues and only minor stylistic issues, so I'll go ahead and push changes for them.

**Review by StephanTLavavej — APPROVED** [review]

File: stl/src/filesystem.cpp
Comment (StephanTLavavej) [inline stl/src/filesystem.cpp:136]: As I mentioned before, we conventionally use direct-initialization here; `Type variable{initializer};` is simpler than `Type variable = Type{initializer};` with no loss of clarity. Precedent:

https://github.com/microsoft/STL/blob/9947dd9d6d2506029ed4486bb7596bda7d70f99d/stl/src/filesystem.cpp#L718

**Review by StephanTLavavej — APPROVED** [review]

File: stl/src/filesystem.cpp
Comment (strega-nil-ms) [inline stl/src/filesystem.cpp:81]:
```suggestion
        if (CopyFileW(_Source, _Target, _Fail_if_exists)) {
```

this change is really weird and not good

**Comment by AtariDreams** [comment]
Isn't that the convention?

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
@strega-nil-ms I pushed a merge with `main`, resolving a trivial merge conflict with #2802's detached comment. Then I audited `filesys.cpp` and `filesystem.cpp` for all comparisons of `BOOL` and `BOOLEAN` against `0` - both pre-existing occurrences, and ones that were introduced by this PR. (I checked every single function, instead of doing a search-and-replace, and was careful to not flip any comparisons accidentally.) I avoided changing the code that #2771 is replacing. I also reversed `DeviceIoControl()`'s flow to be if-success, since that's consistent with other changes here.

@AtariDreams Thanks for your patience - we should be able to merge this soon. The last thing is that we have to avoid significant changes to `msvcp140.dll` until VS 2022 17.4 Preview 2 is open for changes, so after Nicole double-checks the final changes here, we'll need to hold this PR in a `blocked` state for a little while along with a few other PRs we've accumulated.

File: stl/src/filesystem.cpp
Comment (strega-nil-ms) [inline stl/src/filesystem.cpp:599]: I'm still curious why this is using `0u` instead of just `0`; no changes requested.

**Review by strega-nil-ms — APPROVED** [review]
Thanks, approved!

**Comment by strega-nil-ms** [comment]
This is now Ready To Merge, but I'm marking it as Blocked until changes start flowing into VS 2022 17.4 Preview 2 in a couple of weeks.

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for finding and improving this filesystem code! :open_file_folder: :white_check_mark: :smile_cat:

**Pull request merged**
